app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt

```diff
@@ -0,0 +1,430 @@
+      ).toLiveData()
+
+    // use activity as observer because this fragment's view hasn't been created yet.
+    featureViewStateLiveData.observe(
+      activity,
+      object : Observer<AsyncResult<SpotlightViewState>> {
+        override fun onChanged(it: AsyncResult<SpotlightViewState>?) {
+          if (it is AsyncResult.Success) {
+            val viewState = (it.value)
+            if (viewState == SpotlightViewState.SPOTLIGHT_SEEN) {
+              return
+            } else if (viewState == SpotlightViewState.SPOTLIGHT_NOT_SEEN) {
+              createTarget(spotlightTarget)
+              featureViewStateLiveData.removeObserver(this)
+            }
+          }
+        }
+      }
+    )
+  }
+
+  private fun createTarget(
+    spotlightTarget: SpotlightTarget
+  ) {
+
+    val target = Target.Builder()
+      .setShape(getShape(spotlightTarget))
+      .setOverlay(requestOverlayResource(spotlightTarget))
+      .setOnTargetListener(object : OnTargetListener {
+        override fun onStarted() {
+        }
+
+        override fun onEnded() {
+          targetList.pop()
+          val profileId = ProfileId.newBuilder()
+            .setInternalId(internalProfileId)
+            .build()
+          spotlightStateController.markSpotlightViewed(profileId, spotlightTarget.feature)
+        }
+      })
+      .build(spotlightTarget.anchor)
+
+    targetList.add(target)
+    if (!isSpotlightActive) startSpotlight()
+  }
+
+  private fun startSpotlight() {
+    if (targetList.isNullOrEmpty()) return
+    spotlight = Spotlight.Builder(activity)
+      .setTargets(targetList)
+      .setBackgroundColorRes(R.color.spotlightBackground)
+      .setDuration(1000L)
+      .setAnimation(DecelerateInterpolator(2f))
+      .setOnSpotlightListener(object : OnSpotlightListener {
+        override fun onStarted() {
+          isSpotlightActive = true
+        }
+
+        override fun onEnded() {
+          isSpotlightActive = false
+          startSpotlight()
+        }
+      })
+      .build()
+
+    spotlight.start()
+  }
+
+  private fun getShape(spotlightTarget: SpotlightTarget): Shape {
+    return when (spotlightTarget.shape) {
+      SpotlightShape.RoundedRectangle -> {
+        RoundedRectangle(
+          spotlightTarget.anchorHeight.toFloat(),
+          spotlightTarget.anchorWidth.toFloat(),
+          24f
+        )
+      }
+      SpotlightShape.Circle -> {
+        return if (spotlightTarget.anchorHeight > spotlightTarget.anchorWidth) {
+          Circle((spotlightTarget.anchorHeight / 2).toFloat())
+        } else {
+          Circle((spotlightTarget.anchorWidth / 2).toFloat())
+        }
+      }
+    }
+  }
+
+  private fun getArrowWidth(): Float {
+    return this.resources.getDimension(R.dimen.arrow_width)
+  }
+
+  private fun getArrowHeight(): Float {
+    return this.resources.getDimension(R.dimen.arrow_height)
+  }
+
+  private fun getScreenCentreY(): Int {
+    Log.d("overlay screenCentre Y", (screenHeight / 2).toString())
```

Comment:
Please remove the debug lines here & everywhere else in the PR.

Reply:
done